Design API v3

RadWolfie opened the pull request:
Design API v3.0

Closes #10

New features: (see XbSymbolDatabase.h file for functions' details)
* XbSymbolDatabase_GenerateLibraryFilter
* XbSymbolDatabase_GenerateSectionFilter
* XbSymbolDatabase_CreateXbSymbolContext
* XbSymbolContext_ScanLibrary (design for multi-thread safe support)
   * If there is a dependency on another library's symbols. Such as XACTENG library do depend on DSOUND's symbols. Then second or later libraries will not perform the scan until the library is done scanning in the active thread.
* XbSymbolContext_ScanAllLibraryFilter (design for single-threading only)
   * May will change to support multi-thread usage internally.
* XbSymbolContext_RegisterXRefs (currently is a stub function)
  * Can be resolved by #39
* XbSymbolContext_Release

Currently, the design is backward compatibility. Until final stage will move the old APIs into new APIs.

RadWolfie commented:
Currently, multi-thread support is not enabled for MSVC due to the developers haven't implemented C11 support in their compiler. From what I had read in the recent past, they're working on it and should be bundle in future Visual Studio 2019 update.

RadWolfie commented:
Plus most importantly, include pointer to kernel thunk's address to translate internally for make stronger OOVPA signatures easier. 😄

RadWolfie commented:
If the team want to make the final decision for API's names later, please let me know here or on discord. Otherwise, we need to continue discussion what's need to be add/remove/update the parameters and/or API's names.

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -447,6 +834,33 @@ static const char* xbe_type_str[XB_XBE_TYPE_MAX] = {
+    uint32_t        xb_virt_addr;
+    uint32_t        buffer_size;
+    void*           buffer_lower;
+} Section;
```
These very generic names should get prefixes to avoid conflicts with other APIs / user-code.

Ideally this was true (and consistent) for *all* types / variables / functions in the public headers.

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
 void XbSymbolFirstDetectAddressOnly(bool enable);
 
+// ******************************************************************
+// * API v3 (context) functions to use with other projects.
```
Why even keep the old API? People who don't want to use the new API can just check out the old version (at loss of features), or make independent wrappers.
Porting apps is probably easy eitherway (and there aren't many users to begin with).

RadWolfie replied:
Updated in the comment section.

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
+/// <param name="library_input">See XbSymbolGenerate_LibraryFilter for details.</param>
+/// <param name="section_input">See XbSymbolGenerate_SectionFilter for details.</param>
+/// <param name="kernel_thunk">Input decoded virtual kernel thunk.</param>
+bool XbSymbol_Init(XbSymbolContext* ppContext, xb_symbol_register_t register_func, LibraryHeader library_input, SectionHeader section_input, xbaddr kernel_thunk);
```
Why do I `XbSymbol_Init` to *create*?
And why does `Init` happen as step 3?
This implies the name is bad.

RadWolfie replied:
There are necessary steps must be done first before pass on two data arrays. But yeah, I see your point for whole name is bad. It is still possible to manually create both data arrays on your own if using static library method.

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
+/// Step 4: Perform a manual scan to set xreferences manually by requirement.
+/// </summary>
+/// <param name="pContext">Input XbSymbolContext handler.</param>
+void XbSymbolScan_Manual(XbSymbolContext pContext);
```
These new prefixes are weird?
Functions with different prefix interact; but unlike COM, there don't seem to be objects involved.

I'd expect this to be `XbSymbolContext_ScanManual` (`Object_Method`) or `XbSymbolScanManual` (`PrefixMethod` - existing style) or `XbSymbol_ScanManual` ('Module_Method')

RadWolfie replied:
True, I will fix it later on in this pull request.

RadWolfie replied:
fixed

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -447,6 +834,33 @@ static const char* xbe_type_str[XB_XBE_TYPE_MAX] = {
+    Section*        filters;
+} SectionHeader;
+
+typedef void* XbSymbolContext;
```
This should be renamed to suggest that it is a handle or reference.

Current name implies a struct / class / object; so I'd expect code to work like this (which is incorrect):

```c
XbSymbolContext* ctx = malloc(sizeof(XbSymbolContext));
XbSymbol_Init(&XbSymbolContext, ...);
```

or (also incorrect):

```c
XbSymbolContext* ctx = XbSymbol_CreateContext(...);
```

In C, you usually assume a type to be an object that you don't want to assign / pass by value (as it might be too big). So if you have a non-struct-like-type, something should imply that it's a number / pointer of some sort.

RadWolfie replied:
fixed

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
+// before call XbSymbol_Init function.
+
+/// <summary>
+/// Step 1: Generate library array for LibraryHeader input.
```
Ideally the API would imply the order of calls, so we don't have to document it.

RadWolfie replied:
Well, if use a document tool such as doxygen will generate a documentation for human readability as a help/research tool. Beside, if hover above the API name also provide details too.

RadWolfie replied:
Well, if use a document tool such as doxygen will generate a documentation for human readability as a help/research tool. Beside, if hover above the API name also provide details too.

JayFoxRox replied:
I still think this implies bad design.

People can also learn the order by looking at the code sample, or it will be implied by what each function is doing (which implies the order).
For example, you can't possibly do: `XbSymbolContext_ScanManual` if you didn't create an object of type `XbSymbolContext` before (using `XbSymbolContext_Create` - which is a misnomer at the moment). This implies an order without having to explicitly document it.

In doxygen, you wouldn't be putting the step information near each function. Instead, you'd keep the generic documentation (order of calls) separated.

RadWolfie replied:
Right now, there's no sample in this repo. Nor has a tool to for scan xbe files. That's the main reason steps are in the documentations. Later down the road, all of the steps comment will be remove.

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
+void XbSymbolContext_ScanManual(XbSymbolContextHandle pHandle);
+
+/// <summary>
+/// Step 5: (multi-thread) Process individual library input by third-party.
```
What makes this multi-thread?
Is it just thread-safe, or does it actually spawn threads internally?

And why is this part of the "Step" instead of generic information in the summary?

RadWolfie replied:
Multi-thread safe, if the compiler support C11 standard. Currently, msvc compiler doesn't have C11 standard support.

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
+/// <param name="pHandle">Input XbSymbolContextHandle handler.</param>
+/// <param name="pLibrary">Input pointer of a library to start a scan process.</param>
+/// <param name="xref_first_pass">True: Optimized first search scan process; False: 2nd call and later will continue to return non-zero; see return for detail.</param>
+/// <returns>Return total xref count found. If return zero, then there's nothing left to find.</returns>
```
> If return zero, then there's nothing left to find.

This isn't grammatically correct.
Should this be or something like this:

> If it returns zero, then there's nothing left to find.

Could be made even better with more changes.

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
+unsigned int XbSymbolContext_ScanLibrary(XbSymbolContextHandle pHandle, const XbSDBLibrary* pLibrary, bool xref_first_pass);
+
+/// <summary>
+/// Step 5: (single-thread) Process all of the libraries internally.
```
The single-thread note feels unnecessary.

RadWolfie replied:
It's use for awareness, beside I have reword it to understand better.

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
+/// Step 5: (single-thread) Process all of the libraries internally.
+/// </summary>
+/// <param name="pHandle">Input XbSymbolContextHandle handler.</param>
+void XbSymbolContext_ScanFull(XbSymbolContextHandle pHandle);
```
Maybe rename to `ScanAllLibraries` if that's what it does (or does it do anything else, other than scanning libs?).

RadWolfie replied:
It's more of a scan all filter libraries. The proper wording for me would be `ScanAllFilterLibs`.

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
+void XbSymbolContext_ScanFull(XbSymbolContextHandle pHandle);
+
+/// <summary>
+/// Step 6: Register any xreferences may not had been output during the scan process.
```
Note: Sometimes you use xreferences (here) and sometimes xrefs.
For comments, I'd suggest calling them "cross-references" for clarity.

RadWolfie replied:
"x" wasn't used for "cross", it was actually "xbox" internal usage in the past codebase. And it still sort of is, I'll extend the comment about this.

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
+/// <param name="library_input">See XbSymbolDatabase_GenerateLibraryFilter for details.</param>
+/// <param name="section_input">See XbSymbolDatabase_GenerateSectionFilter for details.</param>
+/// <param name="kernel_thunk">Input decoded virtual kernel thunk.</param>
+bool XbSymbolContext_Create(XbSymbolContextHandle* ppHandle, xb_symbol_register_t register_func, XbSDBLibraryHeader library_input, XbSDBSectionHeader section_input, xbaddr kernel_thunk);
```
If you go with the `Object_Action` naming scheme, then this is a misnomer.
You'd want to call this `XbSymbolDatabase_CreateXbSymbolContext` or something (so it creates the object which can be used later to operate on).

RadWolfie replied:
Sorry, some of my changes got lost in the process. It's suppose to be `XbSymbolContextCreate`.

JayFoxRox replied:
Even that name is somewhat unstandard and fragments the API further.

The proposed name would make this the only function that is *not* of style `Object_Action`.
It would also lead to the old issue with overlapping prefixes (`XbSymbol` and `XbSymbolContext`).

If we keep the old API, too (we shouldn't.. but this PR doesn't seem to remove it yet), then this is even more confusing, as that already used such naming conventions.

I'm not sure how a user could figure out `XbSymbolContextCreate` without looking at the headers (and even then it's confusing to me).

As a user, I'd have expected the call I have mentioned (if the rest of the API uses objects or a singleton).

RadWolfie replied:
👌

JayFoxRox commented on XbSymbolDatabase.h:
```diff
@@ -551,6 +951,73 @@ void XbSymbolContinuousSigScan(bool enable);
+/// <param name="xb_header_addr"></param>
+/// <param name="section_out">Input null pointer will provide the count to allocate memory, then allocate memory to section_out.filters field.</param>
+/// <param name="is_raw">True: Convert array to use raw xbe's relative address; False: Convert array to use xbox virtual memory relative address.</param>
+unsigned int XbSymbolDatabase_GenerateSectionFilter(const void* xb_header_addr, XbSDBSectionHeader* section_out, bool is_raw);
```
This stuff (filters) could probably be treated similar to the function which creates a context (returning a handle and share the naming conventions: `Create` or something).

Basically this would imply that you have to create the filters first, in order to be able to create the context (as it depends on the filters). Then you can operate on the context.

So the API will eventually become self-documenting.

RadWolfie replied:
I'm avoiding handle for both header's structures. Since I don't have a valid reason for one time usage which is only used in create process. And it's easier to reduce extra workload for simple allocation and set the new temporary buffer once. Once `XbSymbolContextCreate` have a copy of the filter's structure data. Then it is safe to free temporary buffer. Unless use the alternate function for multi-threading support, then you'll have to provide library info to `XbSymbolContext_ScanLibrary`.

Plus, I don't plan on implement a reference counter system.

RadWolfie commented:
Since the API naming and comments was changed since initialize commit. I do know initialize commit is working and had been verified using my unit test project locally.

RadWolfie commented:
What's been change:

Added:
* XbSymbolDatabase_GetKernelThunkAddress

Renamed:
* XbSymbolContext_RegisterLibrary
* XbSymbolContext_SetBypassBuildVersionLimit
* XbSymbolContext_SetContinuousSigScan
* XbSymbolContext_SetFirstDetectAddressOnly
* XbSymbolDatabase_LibraryVersion
* XbSymbolDatabase_SetOutputMessage
* XbSymbolDatabase_LibraryToFlag
* XbSymbolDatabase_TestOOVPAs
* XbSymbolDatabase_SetOutputVerbosity

---

Internal functions has been renamed in second commit.

Plus added:
* iXbSymbolContext_Lock
* iXbSymbolContext_Unlock
* iXbSymbolContext_AllowSetParameter
* iXbSymbolContext_AllowScanLibrary

RadWolfie commented:
**NOTE:** `XbSymbolScan` function has not been remove yet. It is now deprecated function.

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -84,59 +89,97 @@ static inline uint32_t BitScanReverse(uint32_t value)
+    LT_GRAPHIC,
+    LT_NETWORK,
+    LT_COUNT,
+    LT_UNKNOWN=LT_COUNT
```
It's cleaner to swap these two, so list LT_UNKNOWN first, and declare LT_COUNT as an assignment of LT_UNKNOWN.

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -359,21 +488,190 @@ uint32_t XbSymbolLibraryToFlag(const char* library_name)
+                    }
+                }
+            }
+            count++;
```
This causes the `filters` array to have gaps. If that's undesirable, increase `count` only when populating a filter.

Same happens below, I'll mark them with a single `gap` comment.

RadWolfie replied:
I think I see what you mean, I'll re-review this function again.

RadWolfie replied:
I don't see the problem. I think you had missed the API's documentation. If `library_header` is null pointer, it will return the counts for what is needed to allocate filter's array. Then re-call the function again to fill in the filters.

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -359,21 +488,190 @@ uint32_t XbSymbolLibraryToFlag(const char* library_name)
+                    library_header->filters[count].qfe_version = 0;
+                    memcpy(library_header->filters[count].name, Lib_DSOUND, 8);
+                }
+                count++;
```
`gap`

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -359,21 +488,190 @@ uint32_t XbSymbolLibraryToFlag(const char* library_name)
+                library_header->filters[count].qfe_version = 0;
+                memcpy(library_header->filters[count].name, Lib_JVS, 8);
+            }
+            count++;
```
`gap`

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -359,21 +488,190 @@ uint32_t XbSymbolLibraryToFlag(const char* library_name)
+                        }
+                    }
+
+                    count++;
```
`gap`

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -84,59 +89,97 @@ static inline uint32_t BitScanReverse(uint32_t value)
 SymbolDatabaseList SymbolDBList[] = {
     // Support inline functions in .text section
-    { XbSymbolLib_D3D8 | XbSymbolLib_D3D8LTCG,{ Sec_text, Sec_D3D, Sec_FLASHROM }, D3D8_OOVPA, D3D8_OOVPA_COUNT },
+    { LT_D3D, XbSymbolLib_D3D8 | XbSymbolLib_D3D8LTCG,{ Sec_text, Sec_D3D, Sec_FLASHROM }, D3D8_OOVPA, D3D8_OOVPA_COUNT },
```
Why include a library type enum value here, when there's also a `internal_GetLibraryType` function that's doing the same? Either remove this variable, and use the function, or remove the function (or implement it as a lookup in SymbolDBList, instead of switch/case code)

RadWolfie replied:
Okay, I'll need to take a step back to see what you're seeing. I'm sure your suggestion method is better. 😁

RadWolfie replied:
After reviewing again, obviously this variable is no longer needed. 😃

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -658,39 +956,113 @@ void XbSymbolScanOOVPA(OOVPATable *OovpaTable,
+        case XbSymbolLib_XNETS:
+        case XbSymbolLib_XNETN:
+            return LT_NETWORK;
+    }
```
Remove this function, or implement it as a lookup in `SymbolDBList` (instead of switch/case code).

RadWolfie replied:
Since SymbolDBList could be used two or more times in a single library scan. This function, `internal_GetLibraryType` will stay to keep the scan process active once. 😉

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -762,158 +1134,253 @@ bool XbSymbolScanSection(uint32_t xbe_base_address,
+    memcpy(pContext->library_input.filters, library_input.filters, sizeof(XbSDBLibrary) * library_input.count);
+    memcpy(pContext->section_input.filters, section_input.filters, sizeof(XbSDBSection) * section_input.count);
+
+                            // Mark all Xrefs initially as undetermined
```
nit : indentation

PatrickvL commented on XbSymbolDatabase.h:
```diff
@@ -91,12 +91,393 @@ extern "C" {
+    KT_IMP_FUNC_MmDbgQueryAvailablePages,               // 0x0178 (376) DEVKIT ONLY!
+    KT_IMP_FUNC_MmDbgReleaseAddress,                    // 0x0179 (377) DEVKIT ONLY!
+    KT_IMP_FUNC_MmDbgWriteCheck,                        // 0x017A (378) DEVKIT ONLY!
+    KT_IMP_COUNT,
```
Instead of `KT_IMP_` prefix, use `XREF_KT_` prefix.

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -1430,178 +1869,252 @@ bool XbLibraryScan(custom_scan_func_t custom_scan_func,
+    library_input.count = XbSymbolDatabase_GenerateLibraryFilter(xb_header_addr, NULL);
+    // If total sum is zero, then the input is invalid.
+    if (library_input.count == 0) {
+        goto EmptyCleanup;
```
Avoid goto, just return here.

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -1430,178 +1869,252 @@ bool XbLibraryScan(custom_scan_func_t custom_scan_func,
+    library_input.filters = malloc(sizeof(XbSDBLibrary) * library_input.count);
+    // Verify if memory has been allocated.
+    if (library_input.filters == NULL) {
+        goto EmptyCleanup;
```
Avoid goto, just return here.

PatrickvL commented on XbSymbolDatabase.c:
```diff
@@ -658,39 +956,113 @@ void XbSymbolScanOOVPA(OOVPATable *OovpaTable,
+{
+    xb_xbe_type xbe_type = GetXbeType(xb_header_addr);
+
+    uint32_t kernel_thunk_addr = *(uint32_t*)((uint8_t*)xb_header_addr + 0x0158);
```
Use a symbol instead of the 'magic' literal hexadecimal number `0x0158`.

RadWolfie replied:
I didn't make any modification here except move into its own function.

RadWolfie commented:
Fixed all reviews by @PatrickvL except one which is working as intended.

Plus I discover a warning message from Visual Studio about memcpy usage over the limit. I have replaced memcpy to strncpy function for prevent warning and null'd the overflow.
